orderer/common/cluster/block_fetcher.go

```diff
@@ -0,0 +1,94 @@
+package cluster
+
+import (
+	"github.com/hyperledger/fabric-protos-go/common"
+	"github.com/hyperledger/fabric/bccsp"
+)
+
+//go:generate mockery -dir . -name BlockSource -case underscore -output mocks/
+
+type BlockSource interface {
+	PullBlock(seq uint64) *common.Block
+	UpdateEndpoints(endpoints []EndpointCriteria)
+}
+
+type BlockFetcher struct {
+	source                 BlockSource
+	Endpoints              []EndpointCriteria
+	suspects               []string // a list of bft suepected notes, for optimisation later can be converted to a set.
+	endpointToPull         string
+	endpointUnderSuspicion bool
+}
+
+func (bf *BlockFetcher) isSuspect(suspect string) bool {
+	for _, r := range bf.suspects {
+		if suspect == r {
+			return true
+		}
+	}
+	return false
+}
+
+func (bf *BlockFetcher) chooseEndpoint(ec EndpointCriteria) {
```

Comment:
why `choose` endpoint and not `set` endpoint ? We don't choose here from different endpoints

Reply:
agree